src/ringct/multiexp.cc

```diff
@@ -571,43 +577,33 @@ size_t get_pippenger_c(size_t N)
   return 9;
 }
 
-struct pippenger_cached_data
-{
-  size_t size;
-  ge_cached *cached;
-  pippenger_cached_data(): size(0), cached(NULL) {}
-  ~pippenger_cached_data() { aligned_free(cached); }
-};
-
 std::shared_ptr<pippenger_cached_data> pippenger_init_cache(const std::vector<MultiexpData> &data, size_t start_offset, size_t N)
 {
   MULTIEXP_PERF(PERF_TIMER_START_UNIT(pippenger_init_cache, 1000000));
   CHECK_AND_ASSERT_THROW_MES(start_offset <= data.size(), "Bad cache base data");
   if (N == 0)
     N = data.size() - start_offset;
   CHECK_AND_ASSERT_THROW_MES(N <= data.size() - start_offset, "Bad cache base data");
-  std::shared_ptr<pippenger_cached_data> cache(new pippenger_cached_data());
+  std::shared_ptr<pippenger_cached_data> cache = std::make_shared<pippenger_cached_data>();
 
-  cache->size = N;
-  cache->cached = (ge_cached*)aligned_realloc(cache->cached, N * sizeof(ge_cached), 4096);
-  CHECK_AND_ASSERT_THROW_MES(cache->cached, "Out of memory");
+  cache->resize(N);
```

Comment:
The old code had `CHECK_AND_ASSERT_THROW_MES(cache->cached, "Out of memory");` in case of allocation failure, this new code doesn't handle this case. `std::vector::resize` will throw `std::bad_alloc` on allocation failure, it should be handled here.

Reply:
On the other hand, the only difference will be what type of assertion will be thrown here: `CHECK_AND_ASSERT_THROW_MES` throws `std::runtime_error` instead of `std::bad_alloc`. I couldn't find places in code that catch explicitly `std::runtime_error`, so it shouldn't matter.

Reply:
Actually, I just found `catch (const std::runtime_error& e)` in `zmq_server.cpp` and `simplewallet.cpp`, so it makes sense to catch `std::bad_alloc` here and use the standard `CHECK_AND_ASSERT_THROW_MES`

Reply:
I'm 90% sure this is a `CHECK_AND_ASSERT_THROW_MES()` because A) `aligned_realloc` does not throw an exception and needs to be checked, B) @moneromooo-monero only uses `CHECK_AND_ASSERT_THROW_MES()` and would never write `throw std::bad_alloc;`. I have seen nowhere in the code base where bad allocs are rethrown as runtime errors.

Reply:
`std::bad_alloc` is thrown in RandomX code. The other question is, should it be handled in the code at all, or just let everything crash. Probably the latter because out of memory situations are hard to recover from. I think it's fine to not catch it here.

Reply:
UkoeHB is right, I seldom send separate exceptions. Sometimes std::rnutime_error, that's it. I'm pretty sure I did not write code that relies on std::bad_alloc here (though of course who knows what anyone else might have done later on).